File: backends/php-laravel/app/Http/Controllers/ModerationController.php
<?php

namespace App\Http\Controllers;

use App\Auth\Caller;
use App\Http\Requests\BookmarkStatusRequest;
use App\Http\Requests\ReportRequest;
use App\Http\Requests\ResolutionRequest;
use App\Http\Resources\BookmarkResource;
use App\Http\Resources\ReportResource;
use App\Models\Bookmark;
use App\Models\Report;
use App\Services\AuditService;
use App\Support\BadRequestProblem;
use App\Support\ConflictProblem;
use App\Support\Logger;
use App\Support\NotFoundProblem;
use App\Support\Wire;
use Illuminate\Database\QueryException;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Str;

class ModerationController extends Controller
{
    private const STATUSES = ['open', 'dismissed', 'actioned'];

    public function __construct(private readonly AuditService $audit) {}

    public function reportBookmark(ReportRequest $request, string $id): JsonResponse
    {
        $caller = Caller::require($request);
        $bookmarkId = Wire::parseUuid($id);
        $input = $request->contractData();
        $report = DB::transaction(function () use ($caller, $bookmarkId, $input): Report {
            $bookmark = Bookmark::query()->lockForUpdate()->find($bookmarkId);
            if ($bookmark === null || $bookmark->visibility !== 'public' || $bookmark->status !== 'active') {
                throw new NotFoundProblem;
            }
            if (Report::query()->where('bookmark_id', $bookmarkId)->where('reporter', $caller->username)->where('status', 'open')->exists()) {
                throw new ConflictProblem('You already have an open report on this bookmark.');
            }
            try {
                return Report::create([
                    ...$input,
                    'id' => (string) Str::uuid(),
                    'bookmark_id' => $bookmarkId,
                    'reporter' => $caller->username,
                    'status' => 'open',
                    'created_at' => now(),
                ]);
            } catch (QueryException $error) {
                if (($error->errorInfo[0] ?? null) === '23505') {
                    throw new ConflictProblem('You already have an open report on this bookmark.');
                }
                throw $error;
            }
        });
        Logger::event('info', 'report_created', 'success', 'Report created on a public bookmark', [
            'actor' => $caller->username,
            'resource_type' => 'report',
            'resource_id' => $report->id,
            'bookmark_id' => $bookmarkId,
            'reason' => $report->reason,
        ]);

        return response()->json($this->toReport($report, $request), 201);
    }

    public function listMine(Request $request): array
    {
        $caller = Caller::require($request);
        ['page' => $page, 'size' => $size] = Wire::paging($request);
        $status = $this->validatedStatus(Wire::singleParam($request, 'status'));
        $query = Report::query()->where('reporter', $caller->username);
        if ($status !== null) {
            $query->where('status', $status);
        }
        $total = (clone $query)->count();
        $reports = $query->orderByDesc('created_at')->orderByDesc('id')->skip($page * $size)->take($size)->get();

        return $this->pageOf($reports, $request, $page, $size, $total);
    }

    public function updateMine(ReportRequest $request, string $id): array
    {
        $caller = Caller::require($request);
        $reportId = Wire::parseUuid($id);

        $input = $request->contractData();

        $updated = DB::transaction(function () use ($caller, $reportId, $input): Report {
            $report = $this->ownReport($caller->username, $reportId);
            $this->requireOpen($report);
            $report->update($input);
            Logger::event('info', 'report_updated', 'success', 'Report updated by its reporter', [
                'actor' => $caller->username,
                'resource_type' => 'report',
                'resource_id' => $reportId,
                'bookmark_id' => $report->bookmark_id,
                'reason' => $input['reason'],
            ]);

            return $report->refresh();
        });

        return $this->toReport($updated, $request);
    }

    public function withdrawMine(Request $request, string $id): JsonResponse
    {
        $caller = Caller::require($request);
        $reportId = Wire::parseUuid($id);
        DB::transaction(function () use ($caller, $reportId): void {
            $report = $this->ownReport($caller->username, $reportId);
            $this->requireOpen($report);
            $report->delete();
            Logger::event('info', 'report_withdrawn', 'success', 'Report withdrawn by its reporter', [
                'actor' => $caller->username,
                'resource_type' => 'report',
                'resource_id' => $reportId,
                'bookmark_id' => $report->bookmark_id,
            ]);
        });

        return response()->json(null, 204);
    }

    public function listAdmin(Request $request): array
    {
        Caller::requireRole($request, 'moderator');
        ['page' => $page, 'size' => $size] = Wire::paging($request);
        $status = $this->validatedStatus(Wire::singleParam($request, 'status')) ?? 'open';
        $query = Report::query()->where('status', $status);
        $total = (clone $query)->count();
        $reports = $query->orderBy('created_at')->orderBy('id')->skip($page * $size)->take($size)->get();

        return $this->pageOf($reports, $request, $page, $size, $total);
    }

    public function resolve(ResolutionRequest $request, string $id): array
    {
        $caller = Caller::requireRole($request, 'moderator');
        $reportId = Wire::parseUuid($id);
        $input = $request->contractData();

        return DB::transaction(function () use ($caller, $reportId, $input, $request): array {
            if ($input['resolution'] === 'actioned') {
                $scalar = Report::query()->select('bookmark_id')->find($reportId);
                if ($scalar === null) {
                    throw new NotFoundProblem;
                }
                Bookmark::query()->lockForUpdate()->find($scalar->bookmark_id);
            }
            $report = Report::query()->lockForUpdate()->find($reportId);
            if ($report === null) {
                throw new NotFoundProblem;
            }

            if ($input['resolution'] === 'open') {
                if (Report::query()->where('bookmark_id', $report->bookmark_id)->where('reporter', $report->reporter)->where('status', 'open')->whereKeyNot($reportId)->exists()) {
                    throw new ConflictProblem('The reporter already has another open report on this bookmark.');
                }
                try {
                    $report->update(['status' => 'open', 'resolved_by' => null, 'resolved_at' => null, 'resolution_note' => null]);
                } catch (QueryException $error) {
                    if (($error->errorInfo[0] ?? null) === '23505') {
                        throw new ConflictProblem('The reporter already has another open report on this bookmark.');
                    }
                    throw $error;
                }
                $this->audit->record($caller->username, 'report.reopened', 'report', $reportId, ['bookmarkId' => $report->bookmark_id]);
                Logger::event('info', 'report_reopened', 'success', 'Report re-opened', [
                    'actor' => $caller->username,
                    'resource_type' => 'report',
                    'resource_id' => $reportId,
                    'bookmark_id' => $report->bookmark_id,
                ]);

                return $this->toReport($report->refresh(), $request);
            }

            $resolved = $this->resolveOne($report, $input['resolution'], $caller->username, $input['note'], false);
            if ($input['resolution'] === 'actioned') {
                $this->hideBookmark($caller->username, $report->bookmark_id, $input['note']);
                $siblings = Report::query()->where('bookmark_id', $report->bookmark_id)->where('status', 'open')->whereKeyNot($reportId)->orderBy('id')->lockForUpdate()->get();
                foreach ($siblings as $sibling) {
                    $this->resolveOne($sibling, 'actioned', $caller->username, $input['note'], true);
                }
            }

            return $this->toReport($resolved, $request);
        });
    }

    public function setBookmarkStatus(BookmarkStatusRequest $request, string $id): array
    {
        $caller = Caller::requireRole($request, 'moderator');
        $bookmarkId = Wire::parseUuid($id);
        $input = $request->contractData();

        $bookmark = DB::transaction(function () use ($caller, $bookmarkId, $input): Bookmark {
            $existing = Bookmark::query()->lockForUpdate()->find($bookmarkId);
            if ($existing === null) {
                throw new NotFoundProblem;
            }
            $previousStatus = $existing->status;
            $existing->status = $input['status'];
            $existing->save();
            $this->audit->record($caller->username, 'bookmark.status-changed', 'bookmark', $bookmarkId, [
                'from' => $previousStatus,
                'to' => $input['status'],
                'note' => $input['note'],
            ]);
            Logger::event('info', 'bookmark_status_changed', 'success', 'Bookmark moderation status changed', [
                'actor' => $caller->username,
                'resource_type' => 'bookmark',
                'resource_id' => $bookmarkId,
                'from' => $previousStatus,
                'to' => $input['status'],
            ]);

            return $existing->refresh();
        });

        return $this->toBookmark($bookmark, $request);
    }

    private function ownReport(string $reporter, string $id): Report
    {
        $report = Report::query()->lockForUpdate()->find($id);
        if ($report === null || $report->reporter !== $reporter) {
            throw new NotFoundProblem;
        }

        return $report;
    }

    private function requireOpen(Report $report): void
    {
        if ($report->status !== 'open') {
            throw new ConflictProblem('The report has already been resolved.');
        }
    }

    private function resolveOne(Report $report, string $resolution, string $actor, ?string $note, bool $autoResolved): Report
    {
        $report->update(['status' => $resolution, 'resolved_by' => $actor, 'resolved_at' => now(), 'resolution_note' => $note]);
        $this->audit->record($actor, 'report.resolved', 'report', $report->id, [
            'bookmarkId' => $report->bookmark_id,
            'resolution' => $resolution,
            'note' => $note,
            'autoResolved' => $autoResolved,
        ]);
        Logger::event('info', 'report_resolved', 'success', 'Report resolved', [
            'actor' => $actor,
            'resource_type' => 'report',
            'resource_id' => $report->id,
            'bookmark_id' => $report->bookmark_id,
            'resolution' => $resolution,
            'auto_resolved' => $autoResolved,
        ]);

        return $report->refresh();
    }

    private function hideBookmark(string $actor, string $bookmarkId, ?string $note): void
    {
        $bookmark = Bookmark::find($bookmarkId);
        if ($bookmark === null) {
            throw new NotFoundProblem;
        }
        if ($bookmark->status === 'hidden') {
            return;
        }
        $bookmark->status = 'hidden';
        $bookmark->save();
        $this->audit->record($actor, 'bookmark.status-changed', 'bookmark', $bookmarkId, [
            'from' => 'active',
            'to' => 'hidden',
            'note' => $note,
        ]);
        Logger::event('info', 'bookmark_status_changed', 'success', 'Bookmark hidden by an actioned report', [
            'actor' => $actor,
            'resource_type' => 'bookmark',
            'resource_id' => $bookmarkId,
            'from' => 'active',
            'to' => 'hidden',
        ]);
    }

    private function validatedStatus(?string $value): ?string
    {
        if ($value === null) {
            return null;
        }
        if (! in_array($value, self::STATUSES, true)) {
            throw new BadRequestProblem("unknown status: $value");
        }

        return $value;
    }

    private function toReport(Report $report, Request $request): array
    {
        return (new ReportResource($report))->resolve($request);
    }

    private function pageOf(iterable $reports, Request $request, int $page, int $size, int $totalItems): array
    {
        return [
            'items' => collect($reports)->map(fn (Report $report): array => $this->toReport($report, $request))->values()->all(),
            'page' => $page,
            'size' => $size,
            'totalItems' => $totalItems,
            'totalPages' => (int) ceil($totalItems / $size),
        ];
    }

    private function toBookmark(Bookmark $bookmark, Request $request): array
    {
        return (new BookmarkResource($bookmark))->resolve($request);
    }
}

File: backends/play-scala/app/controllers/ModerationController.scala
package controllers

import models.*
import play.api.libs.json.{JsBoolean, JsString, Json}
import play.api.mvc.{AbstractController, Action, AnyContent, ControllerComponents}
import repositories.{Db, Rows}
import services.{ApiAction, AuditService, AuthService, EventLogger, SqlErrors}
import support.{BookmarkStatusInput, InputJson, ReportInput, ResolutionInput, Responses, Wire}
import support.InputJson.given
import support.Responses.given

import java.sql.{Connection, SQLException}
import java.time.Instant
import java.util.UUID
import javax.inject.{Inject, Singleton}

@Singleton
class ModerationController @Inject() (
    cc: ControllerComponents,
    api: ApiAction,
    db: Db,
    auth: AuthService,
    logger: EventLogger,
    audit: AuditService
) extends AbstractController(cc) {
  import Wire.*

  def createReport(bookmarkId: String): Action[AnyContent] = api { implicit request =>
    val caller = auth.requireCaller(request)
    val uuid = parseUuid(bookmarkId)
    val input = InputJson.read[ReportInput](request)
    val row = db.transaction { conn =>
      val visible = db
        .one(conn, "select visibility, status from bookmarks where id = ? for update", Seq(uuid)) { rs =>
          rs.getString("visibility") == "public" && rs.getString("status") == "active"
        }
        .getOrElse(false)
      if (!visible) throw new NotFoundProblem
      val open = db
        .one(
          conn,
          "select 1 from reports where bookmark_id = ? and reporter = ? and status = 'open'",
          Seq(uuid, caller.username)
        )(_ => true)
        .getOrElse(false)
      if (open) throw new ConflictProblem("You already have an open report on this bookmark.")
      try
        db.returning(
          conn,
          """insert into reports (id, bookmark_id, reporter, reason, comment, status, created_at)
            |values (?, ?, ?, ?, ?, 'open', ?) returning *""".stripMargin,
          Seq(UUID.randomUUID(), uuid, caller.username, input.reason, input.comment, Instant.now())
        )(Rows.report)
      catch {
        case error: SQLException if SqlErrors.state(error).contains("23505") =>
          throw new ConflictProblem("You already have an open report on this bookmark.")
      }
    }
    logger.event(
      "info",
      "report_created",
      "success",
      "Report created on a public bookmark",
      "actor" -> JsString(caller.username),
      "resource_type" -> JsString("report"),
      "resource_id" -> JsString(row.id.toString),
      "bookmark_id" -> JsString(uuid.toString),
      "reason" -> JsString(row.reason)
    )
    Created(Json.toJson(row))
  }

  def listOwnReports: Action[AnyContent] = api { implicit request =>
    val caller = auth.requireCaller(request)
    val (page, size) = paging(request.queryString)
    val status = validatedReportStatus(single(request.queryString, "status"))
    val (where, params) = status match {
      case Some(value) => ("reporter = ? and status = ?", Seq(caller.username, value))
      case None        => ("reporter = ?", Seq(caller.username))
    }
    val payload = db.withConnection { conn =>
      val rows = db.query(
        conn,
        s"select * from reports where $where order by created_at desc, id desc limit ? offset ?",
        params ++ Seq(size, page * size)
      )(Rows.report)
      val total = count(conn, s"select count(*) as count from reports where $where", params)
      Wire.page(rows.map(row => Json.toJson(row)), page, size, total)
    }
    Ok(payload)
  }

  def updateOwnReport(id: String): Action[AnyContent] = api { implicit request =>
    val caller = auth.requireCaller(request)
    val uuid = parseUuid(id)
    val input = InputJson.read[ReportInput](request)
    val row = db.transaction { conn =>
      val report = ownReport(conn, caller.username, uuid)
      requireOpen(report)
      db.returning(
        conn,
        "update reports set reason = ?, comment = ? where id = ? returning *",
        Seq(input.reason, input.comment, uuid)
      )(Rows.report)
    }
    logger.event(
      "info",
      "report_updated",
      "success",
      "Report updated by its reporter",
      "actor" -> JsString(caller.username),
      "resource_type" -> JsString("report"),
      "resource_id" -> JsString(uuid.toString),
      "bookmark_id" -> JsString(row.bookmarkId.toString),
      "reason" -> JsString(row.reason)
    )
    Ok(Json.toJson(row))
  }

  def withdrawOwnReport(id: String): Action[AnyContent] = api { implicit request =>
    val caller = auth.requireCaller(request)
    val uuid = parseUuid(id)
    val bookmarkId = db.transaction { conn =>
      val report = ownReport(conn, caller.username, uuid)
      requireOpen(report)
      db.execute(conn, "delete from reports where id = ?", Seq(uuid))
      report.bookmarkId
    }
    logger.event(
      "info",
      "report_withdrawn",
      "success",
      "Report withdrawn by its reporter",
      "actor" -> JsString(caller.username),
      "resource_type" -> JsString("report"),
      "resource_id" -> JsString(uuid.toString),
      "bookmark_id" -> JsString(bookmarkId.toString)
    )
    NoContent
  }

  def listReports: Action[AnyContent] = api { implicit request =>
    auth.requireRole(request, "moderator")
    val (page, size) = paging(request.queryString)
    val status = validatedReportStatus(single(request.queryString, "status")).getOrElse("open")
    val payload = db.withConnection { conn =>
      val rows = db.query(
        conn,
        "select * from reports where status = ? order by created_at asc, id asc limit ? offset ?",
        Seq(status, size, page * size)
      )(Rows.report)
      val total = count(conn, "select count(*) as count from reports where status = ?", Seq(status))
      Wire.page(rows.map(row => Json.toJson(row)), page, size, total)
    }
    Ok(payload)
  }

  def resolveReport(id: String): Action[AnyContent] = api { implicit request =>
    val caller = auth.requireRole(request, "moderator")
    val uuid = parseUuid(id)
    val input = InputJson.read[ResolutionInput](request)
    val resolution = input.resolution
    val note = input.note
    val row = db.transaction { conn =>
      if (resolution == "actioned") {
        val bookmarkId = db
          .one(conn, "select bookmark_id from reports where id = ?", Seq(uuid))(
            _.getObject("bookmark_id", classOf[UUID])
          )
          .getOrElse(throw new NotFoundProblem)
        db.one(conn, "select id from bookmarks where id = ? for update", Seq(bookmarkId))(_ => true)
      }
      val report = db
        .one(conn, "select * from reports where id = ? for update", Seq(uuid))(Rows.report)
        .getOrElse(throw new NotFoundProblem)
      if (resolution == "open") {
        val conflict = db
          .one(
            conn,
            "select 1 from reports where bookmark_id = ? and reporter = ? and status = 'open' and id <> ?",
            Seq(report.bookmarkId, report.reporter, uuid)
          )(_ => true)
          .getOrElse(false)
        if (conflict) throw new ConflictProblem("The reporter already has another open report on this bookmark.")
        val reopened =
          try
            db.returning(
              conn,
              "update reports set status = 'open', resolved_by = null, resolved_at = null, resolution_note = null where id = ? returning *",
              Seq(uuid)
            )(Rows.report)
          catch {
            case error: SQLException if SqlErrors.state(error).contains("23505") =>
              throw new ConflictProblem("The reporter already has another open report on this bookmark.")
          }
        audit.record(
          conn,
          caller.username,
          "report.reopened",
          "report",
          uuid.toString,
          Json.obj("bookmarkId" -> report.bookmarkId.toString)
        )
        logger.event(
          "info",
          "report_reopened",
          "success",
          "Report re-opened",
          "actor" -> JsString(caller.username),
          "resource_type" -> JsString("report"),
          "resource_id" -> JsString(uuid.toString),
          "bookmark_id" -> JsString(report.bookmarkId.toString)
        )
        reopened
      } else {
        val resolved = resolveOne(conn, report, resolution, caller.username, note, autoResolved = false)
        if (resolution == "actioned") {
          hideBookmark(conn, caller.username, report.bookmarkId, note)
          val siblings = db.query(
            conn,
            "select * from reports where bookmark_id = ? and status = 'open' and id <> ? order by id asc for update",
            Seq(report.bookmarkId, uuid)
          )(Rows.report)
          siblings.foreach(resolveOne(conn, _, "actioned", caller.username, note, autoResolved = true))
        }
        resolved
      }
    }
    Ok(Json.toJson(row))
  }

  def setBookmarkStatus(id: String): Action[AnyContent] = api { implicit request =>
    val caller = auth.requireRole(request, "moderator")
    val uuid = parseUuid(id)
    val input = InputJson.read[BookmarkStatusInput](request)
    val status = input.status
    val note = input.note
    val row = db.transaction { conn =>
      val existing = db
        .one(conn, "select * from bookmarks where id = ? for update", Seq(uuid))(Rows.bookmark)
        .getOrElse(throw new NotFoundProblem)
      val updated = db.returning(
        conn,
        "update bookmarks set status = ?, updated_at = ? where id = ? returning *",
        Seq(status, Instant.now(), uuid)
      )(Rows.bookmark)
      audit.record(
        conn,
        caller.username,
        "bookmark.status-changed",
        "bookmark",
        uuid.toString,
        Json.obj("from" -> existing.status, "to" -> status, "note" -> note)
      )
      logger.event(
        "info",
        "bookmark_status_changed",
        "success",
        "Bookmark moderation status changed",
        "actor" -> JsString(caller.username),
        "resource_type" -> JsString("bookmark"),
        "resource_id" -> JsString(uuid.toString),
        "from" -> JsString(existing.status),
        "to" -> JsString(status)
      )
      updated
    }
    Ok(Json.toJson(row))
  }

  private def count(conn: Connection, sql: String, params: Seq[Any] = Seq.empty): Long =
    db.one(conn, sql, params)(_.getLong("count")).getOrElse(0L)

  private def findBookmark(conn: Connection, id: UUID): Option[BookmarkRow] =
    db.one(conn, "select * from bookmarks where id = ?", Seq(id))(Rows.bookmark)

  private def ownReport(conn: Connection, reporter: String, id: UUID): ReportRow = {
    val row = db
      .one(conn, "select * from reports where id = ? for update", Seq(id))(Rows.report)
      .getOrElse(throw new NotFoundProblem)
    if (row.reporter != reporter) throw new NotFoundProblem
    row
  }

  private def requireOpen(report: ReportRow): Unit =
    if (report.status != "open") throw new ConflictProblem("The report has already been resolved.")

  private def validatedReportStatus(value: Option[String]): Option[String] =
    value.map { status =>
      if (!Seq("open", "dismissed", "actioned").contains(status))
        throw new BadRequestProblem(s"unknown status: $status")
      status
    }

  private def resolveOne(
      conn: Connection,
      report: ReportRow,
      resolution: String,
      actor: String,
      note: Option[String],
      autoResolved: Boolean
  ): ReportRow = {
    val updated = db.returning(
      conn,
      "update reports set status = ?, resolved_by = ?, resolved_at = ?, resolution_note = ? where id = ? returning *",
      Seq(resolution, actor, Instant.now(), note, report.id)
    )(Rows.report)
    audit.record(
      conn,
      actor,
      "report.resolved",
      "report",
      report.id.toString,
      Json.obj(
        "bookmarkId" -> report.bookmarkId.toString,
        "resolution" -> resolution,
        "note" -> note,
        "autoResolved" -> autoResolved
      )
    )
    logger.event(
      "info",
      "report_resolved",
      "success",
      "Report resolved",
      "actor" -> JsString(actor),
      "resource_type" -> JsString("report"),
      "resource_id" -> JsString(report.id.toString),
      "bookmark_id" -> JsString(report.bookmarkId.toString),
      "resolution" -> JsString(resolution),
      "auto_resolved" -> JsBoolean(autoResolved)
    )
    updated
  }

  private def hideBookmark(conn: Connection, actor: String, bookmarkId: UUID, note: Option[String]): Unit = {
    val bookmark = findBookmark(conn, bookmarkId).getOrElse(throw new NotFoundProblem)
    if (bookmark.status == "hidden") return
    db.execute(
      conn,
      "update bookmarks set status = 'hidden', updated_at = ? where id = ?",
      Seq(Instant.now(), bookmarkId)
    )
    audit.record(
      conn,
      actor,
      "bookmark.status-changed",
      "bookmark",
      bookmarkId.toString,
      Json.obj("from" -> "active", "to" -> "hidden", "note" -> note)
    )
    logger.event(
      "info",
      "bookmark_status_changed",
      "success",
      "Bookmark hidden by an actioned report",
      "actor" -> JsString(actor),
      "resource_type" -> JsString("bookmark"),
      "resource_id" -> JsString(bookmarkId.toString),
      "from" -> JsString("active"),
      "to" -> JsString("hidden")
    )
  }
}
